File: LoanCalculator/LoanReducer.swift
//
//  LoanReducer.swift
//  LoanCalculator
//
//  Created by Ivan Tishchenko on 17.12.2025.
//

import Foundation

// MARK: - Loan Reducer
func loanReducer(state: LoanState, action: LoanAction) -> LoanState {
    var newState = state
    
    switch action {
    case .setAmount(let amount):
        newState.amount = amount
        newState.submissionResult = nil
        saveState(newState)
        
    case .setPeriod(let period):
        newState.periodDays = period
        newState.submissionResult = nil
        saveState(newState)
        
    case .submissionStarted:
        newState.isLoading = true
        newState.submissionResult = nil
        
    case .submissionSuccess(let responseId):
        newState.isLoading = false
        newState.submissionResult = .success(responseId: responseId)
        
    case .submissionFailure(let message):
        newState.isLoading = false
        newState.submissionResult = .failure(message: message)
        
    case .dismissResult:
        newState.submissionResult = nil
        
    case .loadSavedState:
        if let savedState = loadState() {
            newState.amount = savedState.amount
            newState.periodDays = savedState.periodDays
        }
        
    case .submitApplication:
        // Handled by middleware
        break
    }
    
    return newState
}

// MARK: - UserDefaults Persistence
private func saveState(_ state: LoanState) {
    UserDefaults.standard.set(state.amount, forKey: "loan_amount")
    UserDefaults.standard.set(state.periodDays, forKey: "loan_period")
}

private func loadState() -> (amount: Double, periodDays: Int)? {
    let amount = UserDefaults.standard.double(forKey: "loan_amount")
    let period = UserDefaults.standard.integer(forKey: "loan_period")
    
    guard amount > 0 else { return nil }
    
    return (amount: amount, periodDays: period == 0 ? 14 : period)
}

File: Android/app/src/main/java/com/loancalculator/redux/LoanReducer.kt
package com.loancalculator.redux

import android.content.SharedPreferences

/**
 * Loan Reducer - pure function that produces new state from action
 */
fun loanReducer(
    state: LoanState,
    action: LoanAction,
    preferences: SharedPreferences? = null
): LoanState {
    return when (action) {
        is LoanAction.SetAmount -> {
            val newState = state.copy(
                amount = action.amount,
                submissionResult = null
            )
            saveState(newState, preferences)
            newState
        }

        is LoanAction.SetPeriod -> {
            val newState = state.copy(
                periodDays = action.period,
                submissionResult = null
            )
            saveState(newState, preferences)
            newState
        }

        is LoanAction.SubmissionStarted -> {
            state.copy(
                isLoading = true,
                submissionResult = null
            )
        }

        is LoanAction.SubmissionSuccess -> {
            state.copy(
                isLoading = false,
                submissionResult = SubmissionResult.Success(action.responseId)
            )
        }

        is LoanAction.SubmissionFailure -> {
            state.copy(
                isLoading = false,
                submissionResult = SubmissionResult.Failure(action.message)
            )
        }

        is LoanAction.DismissResult -> {
            state.copy(submissionResult = null)
        }

        is LoanAction.LoadSavedState -> {
            loadState(preferences) ?: state
        }

        is LoanAction.SubmitApplication -> {
            // Handled by middleware in Store
            state
        }
    }
}

/**
 * Save state to SharedPreferences
 */
private fun saveState(state: LoanState, preferences: SharedPreferences?) {
    preferences?.edit()?.apply {
        putFloat("loan_amount", state.amount.toFloat())
        putInt("loan_period", state.periodDays)
        apply()
    }
}

/**
 * Load state from SharedPreferences
 */
private fun loadState(preferences: SharedPreferences?): LoanState? {
    preferences ?: return null
    
    val amount = preferences.getFloat("loan_amount", 0f).toDouble()
    if (amount == 0.0) return null
    
    val period = preferences.getInt("loan_period", 14)
    
    return LoanState(
        amount = amount,
        periodDays = period
    )
}

